[posix] create shared library with multithread support for FTD and MTD devices

bahubali-bg opened the pull request:
This commit includes,
- Added feature OT_MULTITHREAD, to create shared library for FTD and MTD devices
- Added config OPENTHREAD_POSIX_CONFIG_LOG_OUTPUT_STDOUT_PRINTF,
  to write formatted output to stdout using printf
- Added config OPENTHREAD_POSIX_CONFIG_MULTITHREAD_SUPPORT_ENABLE, provide
  multithread support with shared library to develop IoT applications.
- Shared library provides features as below,
	- Manage ot instance
	- Multithread support
	- Write formatted output to stdout

* How to build shared libraries
./script/cmake-build posix -DOT_MULTITHREAD=ON -DOT_PLATFORM_NETIF=ON

* Tested and verified with ioter tool
https://github.com/Samsung/ioter

bahubali-bg commented:
Hi

We have samsung mail id as primary email in git hub account. So raised the commit using samsung email ids.
We have signed CLA and added google mail id as secondary email into respective git hub account. And then performed rescan to fetch CLA. But still Missing CLA error is coming for below 3 co-authors.
Co-authored-by: Manjunath Neelappa Sataraddi <m.sataraddi@samsung.com>
Co-authored-by: Tirth Master <tirth.master@partner.samsung.com>
Co-authored-by: Vedansh Bhardwaj <v.bhardwaj@samsung.com>

We prefer to raise commit/pull requests with SAMSUNG EMAIL for all co-authors.

So can you please help us to resolve this CLA issue.

Thank you.

jwhui commented:
@bahubali-bg , can you verify that all of the samsung.com email addresses are included in the Google Group you used for establishing the Corporate CLA?

bahubali-bg commented:
Thanks for comments.
Note: This commit is verfied successfully with Ioter. Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle. Ioter tool is open source implementation of the Matter Thread Emulator.
Link: https://github.com/samsung/ioter

abtink commented on src/posix/ot_instance.c:
```diff
@@ -0,0 +1,323 @@
+/*
+ *  Copyright (c) 2018, The OpenThread Authors.
```
```suggestion
 *  Copyright (c) 2023, The OpenThread Authors.
```

bahubali-bg replied:
Updated

abtink commented on src/posix/ot_instance.c:
```diff
@@ -0,0 +1,323 @@
+    otLogLevel       mLogLevel;          ///< Debug level of logging.
+    bool             mPrintRadioVersion; ///< Whether to print radio firmware version.
+    bool             mIsVerbose;         ///< Whether to print log to stderr.
+} PosixConfig;
```
There are some duplicate/repeat definitions and code here in this file as in `posix/main.cpp`.

Would be good to refactor to avoid repeated/duplicate code?

bahubali-bg replied:
Removed repeated/duplicate code. and changes are moved to src/posix/platform/lib_shared.cpp

abtink commented on src/posix/ot_instance.c:
```diff
@@ -0,0 +1,323 @@
+    return i;
+}
+
+void otCreateInstance(void *arg)
```
We generally use `ot` prefix for OT public or platform APIs. 

Are these intended as new (posix-specific) platform APIs?
- If so, suggest having them declared and documented in a posix-specific platform header (similar to  `openthread-system.h`).
- If not, then suggest renaming and not using `ot` prefix?

bahubali-bg replied:
Updated the APIs as per your comments

abtink reviewed (commented):
Thanks for submitting this PR and your contribution. 
Some suggestion below

bukepo commented on src/core/ftd-shared.cmake:
```diff
@@ -0,0 +1,60 @@
+#
```
it seems this is for posix platform only. If so, can we move the CMake file into `src/posix`? You may use get_target_property() to retrieve SOURCES and INCLUDE_DIRECTORIES of openthread-ftd and openthread-mtd.

bahubali-bg replied:
Updated

abtink commented on src/core/ftd-shared.cmake:
```diff
@@ -0,0 +1,60 @@
+)
+
+add_library(openthread-ftd-shared SHARED
+    ${PROJECT_SOURCE_DIR}/src/posix/platform/lib_shared.cpp
```
Not sure it is good idea to have this `ftd-shared.cmake` file in `src/core`? I think a core `cmake` file should not be aware of or access file under different folder `src/posix`?

bahubali-bg replied:
Moved changes to src/posix/platform

abtink commented on src/posix/platform/include/openthread/openthread-system.h:
```diff
@@ -169,6 +169,45 @@ int otSysMainloopPoll(otSysMainloopContext *aMainloop);
+ * @param[in]       aDebug      A pointer to the debug level.
+ *
+ */
+void otSysGetInstance(otInstance **aInstance, pthread_t *aThreadId, const char *aComPort, uint16_t aDebug);
```
Thinking more on this I feel the idea of providing new posix `otSys` APIs which start a `pthread` and/or lock/wait on a mutx may not good to include directly in OT posix platform:
- These are really OS and platform-specific and may be best left for each product/project to decide.
- One of the key goal is that we don't want OT APIs to provide functions of an OS (replace OS).

bahubali-bg replied:
Using openthread shared library, developer can use OT APIs to manage ot instance with multithread support.
Using shared library with multithread support, Iot application can invoke OT APIs directly without handling IPC mechanism in multithread environment. Hence help to improve development time for Iot application.

Note: Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).

The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

bahubali-bg commented on src/posix/platform/CMakeLists.txt:
```diff
@@ -101,6 +101,14 @@ if(OT_ANDROID_NDK)
+    target_compile_definitions(ot-posix-config
+        INTERFACE "OPENTHREAD_POSIX_CONFIG_LIB_SHARED_ENABLE=1"
+    )
+	target_compile_options(ot-posix-config INTERFACE -fPIC)
```
Updated

bahubali-bg commented on src/posix/platform/include/openthread/openthread-system.h:
```diff
@@ -169,6 +169,45 @@ int otSysMainloopPoll(otSysMainloopContext *aMainloop);
  */
 void otSysMainloopProcess(otInstance *aInstance, const otSysMainloopContext *aMainloop);
 
+#if OPENTHREAD_POSIX_CONFIG_LIB_SHARED_ENABLE
```
As these APIs are specific to config OPENTHREAD_POSIX_CONFIG_MULTITHREAD_SUPPORT_ENABLE.  Hence added under flag OPENTHREAD_POSIX_CONFIG_MULTITHREAD_SUPPORT_ENABLE.

bahubali-bg commented on src/posix/platform/openthread-posix-config.h:
```diff
@@ -373,4 +373,24 @@
+ *
+ */
+#if OPENTHREAD_POSIX_CONFIG_LIB_SHARED_ENABLE
+#define OPENTHREAD_POSIX_CONFIG_DISABLE_SYSLOG 1
```
Updated

bahubali-bg commented on src/posix/platform/openthread-posix-config.h:
```diff
@@ -373,4 +373,24 @@
+ * Define to 1 to enable generation of openthread shared libraries.
+ *
+ */
+#ifndef OPENTHREAD_POSIX_CONFIG_LIB_SHARED_ENABLE
```
Added config OPENTHREAD_POSIX_CONFIG_MULTITHREAD_SUPPORT_ENABLE, provide multithread support with shared library to develop IoT applications.

junyong-sim commented on src/core/mtd.cmake:
```diff
@@ -36,6 +36,13 @@ target_compile_options(openthread-mtd PRIVATE
 
+if (OT_SHARED_LIB)
+    target_compile_definitions(openthread-ftd PRIVATE
+        OPENTHREAD_CONFIG_MAX_STATECHANGE_HANDLERS=3
```
target_compile_definitions(openthread-ftd PRIVATE
        OPENTHREAD_CONFIG_MAX_STATECHANGE_HANDLERS=3
-> 
target_compile_definitions(openthread-mtd PRIVATE
        OPENTHREAD_CONFIG_MAX_STATECHANGE_HANDLERS=3

junyong-sim replied:
done

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/pull/9058/commits/c4587576ffa6a9fe06621d52b68ecbc0804802e5

And also updated the comments for your queries/suggestions.

Note: Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/commit/c5a806f59549a0a4fbf66655d2380b2c0899ef7c

And also updated the comments for your queries/suggestions.

Note: 
1. Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

2. All changes are merged in above mentioned commit ID.

abtink commented on src/core/ftd.cmake:
```diff
@@ -36,6 +36,13 @@ target_compile_options(openthread-ftd PRIVATE
+        OPENTHREAD_CONFIG_MAX_STATECHANGE_HANDLERS=3
+    )
+    target_compile_options(openthread-ftd PRIVATE -fPIC)
+endif()
```
Can we move these to `posix` cmake files?

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+    if (!getRadioURL(initParam->comPort))
+    {
+        syslog(LOG_CRIT, "radio device not found");
+        return NULL;
```
Per style guide we should avoid having multiple `return` from same function/method.
Can use `code_utils.hpp` macros `VerifyOrExit()` or `ExitNow()`,

There are other functions in this file with multiple `return`s.

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+    /* For multiple devices and intefaces,
+     * we will align the index of them.
+     */
+    sprintf(radioDevice, "/dev/%s", comPort);
```
Use `snprintf()` to avoid potential overflow of buffer, specially since `comPort` is provided by users?

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+    if (gInstance)
+    {
+        syslog(LOG_INFO, "ot instance already initialised!!!");
+        *instance = gInstance;
```
Who `free()` the allocated `initParam` if we exit here?

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+    if (pthread_mutex_init(&gLock, NULL) != 0)
+    {
+        syslog(LOG_INFO, "mutex init has failed");
+        return;
```
Same here? allocated `initParm` can leak?

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+void otSysGetInstance(otInstance **instance, const char *comPort, uint16_t debug)
+{
+    Param *initParam = (Param *)malloc(sizeof(Param));
+    strcpy(initParam->comPort, comPort);
```
Use `strncpy()` to be safe and avoid any overwrite over `initParm->comPort` buffer size?

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+typedef struct Param
+{
+    char     comPort[MAX_TTY_LEN];
+    uint16_t debugLevel;
```
Use `mComPort` instead of `comPort` (to follow style guide).

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+        syslog(LOG_CRIT, "radio device not found");
+        return NULL;
+    }
+    sprintf(radioUrl, "spinel+hdlc+uart:///dev/%s", initParam->comPort);
```
Use `snprintf()` to be safe

bahubali-bg replied:
Updated

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+
+    for (i = 0; i < MULTIPLE_INSTANCE_MAX; i++)
+    {
+        sprintf(iface, "/sys/class/net/wpan%d", i);
```
Use `snprintf()` to be safe?

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+    syslog(LOG_INFO, "Inside pThreadOtMainLoop");
+
+    interfaceIdx = getInterface();
+    sprintf(iface, "wpan%d", interfaceIdx);
```
Use `snprintf()?

abtink commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+
+    pthread_mutex_unlock(&gLock);
+
+    return;
```
remove `return`?

abtink reviewed (commented):
Thanks for update. 
Some suggestions below

junyong-sim commented on src/posix/platform/CMakeLists.txt:
```diff
@@ -192,3 +192,17 @@ target_include_directories(ot-posix-test-settings
+option(OT_SHARED_LIB "Enable generation of openthread shared libraries" OFF)
+if(OT_SHARED_LIB)
+    target_compile_definitions(ot-posix-config
+        INTERFACE "OPENTHREAD_POSIX_CONFIG_MULTITHREAD_SUPPORT_ENABLE=1"
```
I agree your opinion is reasonable. We will discuss about it, and then update you.

junyong-sim commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+    pthread_mutex_lock(&gLock);
+
+    *instance = gInstance;
+#if defined(__linux)
```
I remember this line caused build error on mac os. Anyway, I think we are able to remove this line.

junyong-sim commented on src/posix/platform/shared_lib.cpp:
```diff
@@ -0,0 +1,272 @@
+    pthread_create(&gThreadId, NULL, pThreadOtMainLoop, initParam);
+
+    syslog(LOG_INFO, "Wait for 1 Seconds to initialise openthread stack !!!");
+    sleep(1);
```
I think we can replace sleep with pthread condition variable.
int pthread_cond_wait(pthread_cond_t *restrict cond,
pthread_mutex_t *restrict mutex);

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/pull/9058/commits/a7b43e51661ebc18f913fb8ba971f71f9079e5cc

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

zhongzain commented on src/posix/platform/multithread_api.cpp:
```diff
@@ -0,0 +1,260 @@
+    }
+
+exit:
+    otSysDeinit();
```
Please, call otSysDeinit() only when otSysInit is completed.
```
    if (gInstance)
        otSysDeinit();
```

zhongzain commented on src/posix/platform/multithread_api.cpp:
```diff
@@ -0,0 +1,260 @@
+
+    *aInstance = gInstance;
+
+    otLogDebgPlat("otThread : gThreadId[%ld]", gThreadId);
```
If `otSysInit` is failed, this log will not printed. 
So, please, use `syslog`, instead of `otLogDebgPlat`.
-> 
`syslog(LOG_INFO, "otThread : gThreadId[%ld]", gThreadId);`

zhongzain commented on src/posix/platform/multithread_api.cpp:
```diff
@@ -0,0 +1,260 @@
+    otSysDeinit();
+    gThreadId  = 0;
+    gInstance  = NULL;
+    gTerminate = false;
```
In case mutex is not released because of VerifyOrExit, we should send signal and unlock mutex.
So, I think we have to add following lines here. 
```
    pthread_cond_signal(&gCond);
    pthread_mutex_unlock(&gLock);
```

junyong-sim commented on src/posix/platform/multithread_api.cpp:
```diff
@@ -0,0 +1,260 @@
+    }
+
+exit:
+    otSysDeinit();
```
The `otSysDeinit()` API calls assert when gInstance is null pointer.
So, we have to check if gInstance is not null before calling `otSysDeinit()` 
```
    if (gInstance)
        otSysDeinit();
```

bahubali-bg replied:
Updated

junyong-sim commented on src/posix/platform/multithread_api.cpp:
```diff
@@ -0,0 +1,260 @@
+    otSysDeinit();
+    gThreadId  = 0;
+    gInstance  = NULL;
+    gTerminate = false;
```
If the mutex is not released because of `VerifyOrExit`, we should send signal and unlock the mutex here.
So, I think we have to add following lines.
```
    pthread_cond_signal(&gCond);
    pthread_mutex_unlock(&gLock);
```

bahubali-bg replied:
Updated

junyong-sim commented on src/posix/platform/multithread_api.cpp:
```diff
@@ -0,0 +1,260 @@
+
+    *aInstance = gInstance;
+
+    otLogDebgPlat("otThread : gThreadId[%ld]", gThreadId);
```
If `otSysInit` fails, this log will not printed.
How about we use `syslog`, instead of `otLogDebgPlat` ?
->
syslog(LOG_INFO, "otThread : gThreadId[%ld]", gThreadId);

bahubali-bg replied:
Updated

junyong-sim commented on src/posix/platform/CMakeLists.txt:
```diff
@@ -157,17 +157,17 @@ if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux" AND NOT OT_TARGET_OPENWRT)
+        PRIVATE
+            OPENTHREAD_FTD=1
+            OPENTHREAD_MTD=0
+    )
```
removed

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/pull/9058/commits/bdf4626447dceb273ef456c932439983151fca82

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/commit/5108a1d18cc0df6e589a80fde83e38b3a9c30796

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/pull/9058/commits/5eb74240d845aa9bd5aaeaf43a29507111b21c32

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/commit/f64c1a1a6b7fa683d6f1e12d7942401a38d9c26e

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

abtink commented on src/posix/platform/logging.cpp:
```diff
@@ -69,7 +75,46 @@ OT_TOOL_WEAK void otPlatLog(otLogLevel aLogLevel, otLogRegion aLogRegion, const
     va_start(args, aFormat);
+#if OPENTHREAD_POSIX_CONFIG_LOG_OUTPUT_STDOUT_PRINTF
+    const uint16_t kBufferSize = 4096;
+    char           buffer[kBufferSize];
```
Per style guide, we should define variable at top of scope.
Suggest adding this whole block in it is `{ }`.
```suggestion
#if OPENTHREAD_POSIX_CONFIG_LOG_OUTPUT_STDOUT_PRINTF
    {
         const uint16_t kBufferSize = 4096;
         char           buffer[kBufferSize];
         ...
         
         
    }
```

bahubali-bg replied:
Updated

abtink commented on src/posix/platform/logging.cpp:
```diff
@@ -69,7 +75,46 @@ OT_TOOL_WEAK void otPlatLog(otLogLevel aLogLevel, otLogRegion aLogRegion, const
+#if defined(__APPLE__)
+        int ret = snprintf(buf, sizeof buf, "%s.%06d", tmbuf, tv.tv_usec);
+#else
+        int ret = snprintf(buf, sizeof buf, "%s.%06ld", tmbuf, tv.tv_usec);
```
Same here, `ret` should be defined at the top of scope (after `{`)

bahubali-bg replied:
Updated

abtink commented on src/posix/platform/logging.cpp:
```diff
@@ -69,7 +75,46 @@ OT_TOOL_WEAK void otPlatLog(otLogLevel aLogLevel, otLogRegion aLogRegion, const
+    }
+#else
     vsyslog(aLogLevel, aFormat, args);
+#endif
```
Per style guide if `#endif` is more than 20 lines from related `#if` we need to add comment.

bahubali-bg replied:
Updated

abtink commented on src/posix/platform/openthread-posix-config.h:
```diff
@@ -383,4 +383,24 @@
+ * Define as 1 to enable formatted output to stdout.
+ *
+ */
+#if OPENTHREAD_POSIX_CONFIG_MULTITHREAD_SUPPORT_ENABLE
+#define OPENTHREAD_POSIX_CONFIG_LOG_OUTPUT_STDOUT_PRINTF 1
+#endif
```
We should treat `OPENTHREAD_POSIX_CONFIG_LOG_OUTPUT_STDOUT_PRINTF` as a config and do a `#ifndef` for it before trying to `#define` it.

If you want it to be enabled when  `MULTITHREAD_SUPPORT_ENABLE` is used, we can define it to have the same value:

```suggestion
#ifndef  OPENTHREAD_POSIX_CONFIG_LOG_OUTPUT_STDOUT_PRINTF 
#define OPENTHREAD_POSIX_CONFIG_LOG_OUTPUT_STDOUT_PRINTF OPENTHREAD_POSIX_CONFIG_MULTITHREAD_SUPPORT_ENABLE
#endif
```

bahubali-bg replied:
updated

abtink commented on src/posix/platform/include/openthread/openthread-system.h:
```diff
@@ -242,6 +242,44 @@ typedef struct otSysInfraNetIfAddressCounters
+ * @param[in]    aOtLogLevel  Ot log level.
+ *
+ */
+void otSysGetInstance(otInstance **aInstance, const char *aComPort, uint16_t aOtLogLevel);
```
Use `otLogLevel aLogLevel` instead of `uint16_t aOtLogLevel`?
Need to `#include <openthread/platform/logging.h>` to get the `otLogLevel` definition.

bahubali-bg replied:
Updated

abtink commented on src/posix/platform/multithread_api.cpp:
```diff
@@ -0,0 +1,265 @@
+    VerifyOrExit(pthread_cond_init(&gCond, NULL) == 0, syslog(LOG_INFO, "cond init has failed"));
+
+    pthread_create(&gThreadId, NULL, pThreadOtMainLoop, config);
+    pthread_mutex_lock(&gLock);
```
I think we need to set `config = null` here after passing it `pthread_create ()` so to not `free()` it at exit?

junyong-sim replied:
In old commit, we passed the config to `pThreadOtMainLoop` and then the memory of config was released in `pThreadOtMainLoop`. 
But, in latest commit, this logic has changed.
`pThreadOtMainLoop` uses the config passed and after that, it sends the signal to the caller of pthread_create. 
There is no `free()` in `pThreadOtMainLoop` now.
After getting this signal, `otSysGetInstance` releases the memory of config at `exit`.

So, I think we should not set `config = null` here. 
How about your opinion ?

abtink replied:
I see. got it. Sounds good. Thanks.

abtink reviewed (commented):
Thanks for updating the PR.
Looks good overall. Some smaller suggestions below:

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/pull/9058/commits/feff9541811905918dbd972307be0bf928333e07

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

bahubali-bg reviewed (commented):
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/commit/feff9541811905918dbd972307be0bf928333e07

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

bahubali-bg commented:
Thank for the comments,

We have updated the code as per review comments in the below mentioned commit ID.
https://github.com/openthread/openthread/pull/9058/commits/35126edcb6e0464982a85c174900047a7c16cdf7

And also updated the comments for your queries/suggestions.

Note:

Commit changes were tested and verified with ioter tool (https://github.com/Samsung/ioter).
The Ioter is a tool that allows you to emulate all Matter supported IoT devices with Linux PC and Thread RCP dongle.

abtink reviewed (approved):
LGTM. Thanks.

junyong-sim commented:
@jwhui 
I have found one fail from the result of some checks. I wonder whether or not I have to solve this issue because it seems not to be related to this PR. If I am wrong, please, let me know.
```
======================================================================
ERROR: test (__main__.ManualMulticastAddressConfig)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/thread_cert.py", line 313, in tearDown
    self._verify_packets(test_info_path)
  File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/thread_cert.py", line 342, in _verify_packets
    self.verify(pv)
  File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/border_router/test_manual_maddress.py", line 102, in verify
    _pkt = pkts.filter_eth_src(vars['Host_ETH']) \
  File "/home/runner/work/openthread/openthread/tests/scripts/thread-cert/pktverify/packet_filter.py", line 234, in must_next
    raise errors.PacketNotFound(self.index, self._stop_index)
pktverify.errors.PacketNotFound: ((0, 0), (527, 527))

----------------------------------------------------------------------
Ran 1 test in 88.262s

FAILED (errors=1)
File "./tests/scripts/thread-cert/run_cert_suite.py", line 84, in run_cert
2023-06-30 22:21:35,963 - ERROR - Run test test_manual_maddress_0 failed, please check the log file: ot_testing/test_manual_maddress_0.log
Traceback (most recent call last):
  File "./tests/scripts/thread-cert/run_cert_suite.py", line 76, in run_cert
    subprocess.check_call(abs_script,
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '/home/runner/work/openthread/openthread/tests/scripts/thread-cert/border_router/test_manual_maddress.py' returned non-zero exit status 1.
PASS 0 FAIL 1 ./tests/scripts/thread-cert/border_router/test_manual_maddress.py
Running test_mdns_restart_0
```

jwhui commented:
Closing stale PR.
